parl/framework/computation_task.py

```diff
@@ -0,0 +1,180 @@
+# limitations under the License.
+
+import paddle.fluid as fluid
+import parl.layers as layers
+from parl.framework.algorithm import Model, Algorithm
+
+
+def split_list(l, sizes):
+    """
+    Split a list into several chunks, each chunk with a size in sizes
+    """
+    chunks = []
+    offset = 0
+    for size in sizes:
+        chunks.append(l[offset:offset + size])
+        offset += size
+    return chunks
+
+
+class ComputationTask(object):
+    """
+    A ComputationTask is responsible for the general data flow
+    outside the algorithm
+
+    A ComputationTask is created in a bottom-up way:
+    a. create a Model
+    b. create an Algorithm with the model as an input
+    c. define a ComputationTask with the algorithm
+    """
+
+    def __init__(self, algorithm):
+        assert isinstance(algorithm, Algorithm)
+        self.alg = algorithm
+        ## create an Fluid executor
+        self._define_program()
+        place = fluid.CPUPlace() if self.alg.gpu_id < 0 \
+                else fluid.CUDAPlace(self.alg.gpu_id)
+        self.fluid_executor = fluid.Executor(place)
+        self.fluid_executor.run(fluid.default_startup_program())
+
+    def _create_data_layers(self, specs):
+        data_layers = {}
+        for name, args in specs:
+            data_layers[name] = layers.data(name, **args)
+        return data_layers
+
+    def _define_program(self):
+        self.learn_program = fluid.Program()
+        self.predict_program = fluid.Program()
+
+        def _get_next_specs(specs):
+            return [("next_" + spec[0], spec[1]) for spec in specs]
+
+        def _select_data(data_layer_dict, specs):
+            return {name: data_layer_dict[name] for name, _ in specs}
+
+        input_specs = self.alg.get_input_specs()
+        state_specs = self.alg.get_state_specs()
+        next_input_specs = _get_next_specs(input_specs)
+        next_state_specs = _get_next_specs(state_specs)
+        action_specs = self.alg.get_action_specs()
+        reward_specs = self.alg.get_reward_specs()
+        episode_end_specs = [("episode_end", dict(shape=[1]))]
+
+        self.action_names = sorted([name for name, _ in action_specs])
+        self.state_names = sorted([name for name, _ in state_specs])
+
+        with fluid.program_guard(self.predict_program):
+            data_layer_dict = self._create_data_layers(input_specs)
+            data_layer_dict.update(self._create_data_layers(state_specs))
+            self.predict_feed_names = sorted(data_layer_dict.keys())
+
+            inputs = _select_data(data_layer_dict, input_specs)
+            states = _select_data(data_layer_dict, state_specs)
+
+            ### call alg predict()
+            pred_actions, pred_states = self.alg.predict(inputs, states)
+            self.predict_fetch = [pred_actions, pred_states]
+
+        with fluid.program_guard(self.learn_program):
+            data_layer_dict = self._create_data_layers(input_specs)
```

Comment:
Because get_input_specs can be used in other places. For example, check the consistency between the predict return and the specs to make sure that the user returns correct things.

Reply:
using layers.data seems more intuitive? 
Because users don't have to learn additional api such as _create_data_layers and _select_data.
Can we check the consistency between the predict return and inputs when using layers.data ?

Reply:
_create_data_layers and _select_data are not supposed to be known by the user. This part of code is basically fixed. Users only need to specify get_***_specs(). What I mean by saying checking the consistency between the predict return and the specs is that we expect the user to write a predict function so that it returns several fields whose names can be matched to the return of get_action_specs() and get_state_specs().